backend/FwLite/MiniLcm.Tests/DiffCollectionTests.cs

```diff
@@ -0,0 +1,291 @@
+    [Theory]
+    [MemberData(nameof(CollectionDiffTestCaseData))]
+    public async Task DiffTests(CollectionDiffTestCase testCase)
+    {
+        // act
+        var (changeCount, diffOperations, replacements) = await Diff(testCase.OldValues, testCase.NewValues);
+
+        // assert
+        changeCount.Should().Be(testCase.ExpectedOperations.Count);
+        diffOperations.Should().BeEquivalentTo(testCase.ExpectedOperations, options => options.WithStrictOrdering());
+    }
+
+    private static async Task<(
+        int changeCount,
+        List<CollectionDiffOperation> DiffOperations,
+        List<(TestOrderable before, TestOrderable after)> Replacements
+    )> Diff(TestOrderable[] oldValues, TestOrderable[] newValues)
+    {
+        var diffApi = new TestOrderableDiffApi(oldValues);
+        var changeCount = await DiffCollection.DiffOrderable(oldValues, newValues, diffApi);
+        diffApi.Current.Should().BeEquivalentTo(newValues);
+        var expectedReplacements = oldValues.Join(newValues, o => o.Id, o => o.Id, (o, n) => (o, n)).ToList();
+        diffApi.Replacements.Should().BeEquivalentTo(expectedReplacements);
+        return (changeCount, diffApi.DiffOperations, diffApi.Replacements);
+    }
+
+    private static TestOrderable Orderable(double order, Guid? id = null)
+    {
+        return new TestOrderable()
+        {
+            Order = order,
+            Id = id ?? Guid.NewGuid(),
+        };
+    }
+
+    private static BetweenPosition Between(TestOrderable? previous, TestOrderable? next)
+    {
+        return Between(previous?.Id, next?.Id);
+    }
+
+    private static BetweenPosition Between(Guid? previous = null, Guid? next = null)
+    {
+        return new BetweenPosition
+        {
+            Previous = previous,
+            Next = next
+        };
+    }
+
+    private static CollectionDiffOperation Move(TestOrderable value, BetweenPosition between)
+    {
+        return new CollectionDiffOperation(value, PositionDiffKind.Move, between);
+    }
+
+    private static CollectionDiffOperation Add(TestOrderable value, BetweenPosition between)
+    {
+        return new CollectionDiffOperation(value, PositionDiffKind.Add, between);
+    }
+
+    private static CollectionDiffOperation Remove(TestOrderable value)
+    {
+        return new CollectionDiffOperation(value, PositionDiffKind.Remove);
+    }
+}
+
+public class CollectionDiffTestCase
+{
+    public required TestOrderable[] OldValues { get; init; }
+    public required TestOrderable[] NewValues { get; init; }
+    public required List<CollectionDiffOperation> ExpectedOperations { get; init; }
+}
+
+public record CollectionDiffOperation(TestOrderable Value, PositionDiffKind Kind, BetweenPosition? Between = null);
+
+public class TestOrderable : IOrderable
+{
+    public required Guid Id { get; set; }
+    public required double Order { get; set; }
+}
+
+public class TestOrderableDiffApi(TestOrderable[] before) : OrderableCollectionDiffApi<TestOrderable>
```

Comment:
I kinda feel like since this has so much code in it that it might be worth writing a few tests directly on this, eg `Add` results in x operations. Then I'd probably break it out into it's own file